src/charm.py

```diff
@@ -0,0 +1,151 @@
+        self.framework.observe(self.on.jenkins_pebble_ready, self._on_jenkins_pebble_ready)
+
+    def _get_pebble_layer(self, env: dict[str, str] | None = None) -> Layer:
+        """Return a dictionary representing a Pebble layer.
+
+        Args:
+            env: Map of Jenkins environment variables.
+
+        Returns:
+            The pebble layer defining Jenkins service layer.
+        """
+        default_env = {"JENKINS_HOME": str(JENKINS_HOME)}
+        merged_env = default_env | env if env else default_env
+        layer: LayerDict = {
+            "summary": "jenkins layer",
+            "description": "pebble config layer for jenkins",
+            "services": {
+                "jenkins": {
+                    "override": "replace",
+                    "summary": "jenkins",
+                    "command": "java -Djava.awt.headless=true -jar /srv/jenkins/jenkins.war",
+                    "startup": "enabled",
+                    "environment": merged_env,
+                }
+            },
+        }
+        return Layer(layer)
+
+    def _is_jenkins_ready(self) -> bool:
+        """Check if Jenkins webserver is ready.
+
+        Returns:
+            True if Jenkins server is online. False otherwise.
+        """
+        try:
+            requests.get("http://localhost:8080/login", timeout=10).raise_for_status()
+            return True
+        except requests.HTTPError:
+            return False
+
+    def _wait_jenkins_ready(self, timeout: int = 140, check_interval: int = 10) -> None:
+        """Wait until Jenkins service is up.
+
+        Args:
+            timeout: Time in seconds to wait for jenkins to become ready in 10 second intervals.
+            check_interval: Time in seconds to wait between ready checks.
+
+        Raises:
+            TimeoutError: if Jenkins status check did not pass within the timeout duration.
+        """
+        for _ in range(timeout // check_interval):
+            if self._is_jenkins_ready():
+                break
+            sleep(check_interval)
+        else:
+            raise TimeoutError("Timed out waiting for Jenkins to become ready.")
+
+    def _get_admin_credentials(self, container: Container) -> Credentials:
+        """Retrieve admin credentials.
+
+        Args:
+            container: The Jenkins container.
+
+        Returns:
+            The Jenkins admin account credentials.
+        """
+        user = "admin"
+        password = container.pull(INITIAL_PASSWORD, encoding="utf-8").read().strip()
+        return Credentials(username=user, password=str(password))
+
+    def _unlock_jenkins(self, container: Container) -> None:
+        """Write to executed version and updated version file to bypass Jenkins setup wizard.
+
+        Args:
+            container: The Jenkins container.
+        """
+        credentials = self._get_admin_credentials(container)
+        client = Jenkins("http://localhost:8080", credentials.username, credentials.password)
```

Comment:
Can you move `http://localhost:8080` to a constant and replace all occurrences?

Reply:
Updated, thank you!